zcash_primitives/src/sapling/note_encryption.rs

```diff
@@ -93,235 +73,236 @@ pub fn prf_ock(
-/// let cmu = note.cmu();
-///
-/// let mut enc = SaplingNoteEncryption::new(ovk, note, to, MemoBytes::empty(), &mut rng);
-/// let encCiphertext = enc.encrypt_note_plaintext();
-/// let outCiphertext = enc.encrypt_outgoing_plaintext(&cv.commitment().into(), &cmu);
-/// ```
-pub struct SaplingNoteEncryption<R: RngCore> {
-    epk: jubjub::SubgroupPoint,
-    esk: jubjub::Fr,
-    note: Note,
-    to: PaymentAddress,
-    memo: MemoBytes,
-    /// `None` represents the `ovk = ⊥` case.
-    ovk: Option<OutgoingViewingKey>,
-    rng: R,
-}
+fn sapling_parse_note_plaintext_without_memo<F, P: consensus::Parameters>(
+    domain: &SaplingDomain<P>,
+    plaintext: &[u8],
+    get_validated_pk_d: F,
+) -> Option<(Note, PaymentAddress)>
+where
+    F: FnOnce(&Diversifier) -> Option<jubjub::SubgroupPoint>,
+{
+    assert!(plaintext.len() >= COMPACT_NOTE_SIZE);
 
-impl<R: RngCore + CryptoRng> SaplingNoteEncryption<R> {
-    /// Creates a new encryption context for the given note.
-    ///
-    /// Setting `ovk` to `None` represents the `ovk = ⊥` case, where the note cannot be
-    /// recovered by the sender.
-    pub fn new(
-        ovk: Option<OutgoingViewingKey>,
-        note: Note,
-        to: PaymentAddress,
-        memo: MemoBytes,
-        rng: R,
-    ) -> Self {
-        Self::new_internal(ovk, note, to, memo, rng)
+    // Check note plaintext version
+    if !plaintext_version_is_valid(&domain.params, domain.height, plaintext[0]) {
+        return None;
     }
+
+    // The unwraps below are guaranteed to succeed by the assertion above
+    let diversifier = Diversifier(plaintext[1..12].try_into().unwrap());
+    let value = Amount::from_u64_le_bytes(plaintext[12..20].try_into().unwrap()).ok()?;
+    let r: [u8; 32] = plaintext[20..COMPACT_NOTE_SIZE].try_into().unwrap();
+
+    let rseed = if plaintext[0] == 0x01 {
+        let rcm = jubjub::Fr::from_repr(r)?;
+        Rseed::BeforeZip212(rcm)
+    } else {
+        Rseed::AfterZip212(r)
+    };
+
+    let pk_d = get_validated_pk_d(&diversifier)?;
+
+    let to = PaymentAddress::from_parts(diversifier, pk_d)?;
+    let note = to.create_note(value.into(), rseed)?;
+    Some((note, to))
 }
 
-impl<R: RngCore> SaplingNoteEncryption<R> {
-    pub(crate) fn new_internal(
-        ovk: Option<OutgoingViewingKey>,
-        note: Note,
-        to: PaymentAddress,
-        memo: MemoBytes,
-        mut rng: R,
-    ) -> Self {
-        let esk = note.generate_or_derive_esk_internal(&mut rng);
-        let epk = note.g_d * esk;
-
-        SaplingNoteEncryption {
-            epk,
-            esk,
-            note,
-            to,
-            memo,
-            ovk,
-            rng,
-        }
+pub struct SaplingDomain<P: consensus::Parameters> {
+    params: P,
+    height: BlockHeight,
+}
+
+impl<P: consensus::Parameters> Domain for SaplingDomain<P> {
+    type EphemeralSecretKey = jubjub::Scalar;
+    type EphemeralPublicKey = jubjub::ExtendedPoint;
```

Comment:
This needs to be a byte array, because activation of ZIP 216 in NU5, it can be a non-canonical encoding.